### PR TITLE
download control details

### LINKED ISSUE #4027: Missing thumbnail for pending image download
If an **image** download is pending
there is only the download arrow (and maybe a text)
but no indication (thumbnail) of what will be downloaded.

For audio and video there are the specific built in icons/thumbnails. So maybe we should have that for images too.
Alternative would be to have an ultra low resolution image that is sent with the message as thumbnail.

### PR DESCRIPTION
**Pull request by mcginty**
Show the human-readable content type to the user when the download button appears (eventually showing size too assuming it will be added into the protobuf).

Also fun animations.

fixes #4027

transfercontrols

### CODE REVIEW

**Comment by McLoo** [comment]
Will there also be a an optical refresh to the ~~ugly~~ old style "DOWNLOAD" button for MMS?
(That will hide, once the MMS expired?)

**Comment by McLoo** [comment]
screenshot_2015-09-11-10-39-34 01

**Comment by mcginty** [comment]
@McLoo yeah, I'd like to merge those but I think that's a separate PR's worth of diff :).

File: src/org/thoughtcrime/securesms/components/AnimatingToggle.java [hunk 1]
```diff
@@ -60,6 +60,10 @@ protected void display(@Nullable View view, boolean animated) {
     current = view;
   }
 
+  protected @Nullable View getCurrent() {
+    return current;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/components/AnimatingToggle.java]: this shouldn't be exposed

File: src/org/thoughtcrime/securesms/components/TransferControlView.java [hunk 2]
```diff
@@ -51,25 +69,88 @@ public void setSlide(final @NonNull Slide slide) {
...
+      downloadDetails.setText(slide.getContentDescription());
+      animateIn(downloadDetails);
     } else {
+      setVisibility(GONE);
       display(null, false);
     }
   }
 
+  private void animateIn(@NonNull final View view) {
+    showSmooth();
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/components/TransferControlView.java]: some serious side effects

File: src/org/thoughtcrime/securesms/components/TransferControlView.java [hunk 3]
```diff
@@ -79,7 +160,10 @@ public void onEventAsync(final PartProgressEvent event) {
       Util.runOnMain(new Runnable() {
         @Override public void run() {
           progressWheel.setInstantProgress(((float)event.progress) / event.total);
-          if (event.progress >= event.total) display(null);
+          if (event.progress >= event.total) {
+            hideSmooth();
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/components/TransferControlView.java]: side effect city

File: src/org/thoughtcrime/securesms/mms/Slide.java [hunk 4]
```diff
@@ -63,6 +63,8 @@ public boolean hasAudio() {
     return false;
   }
 
+  public @NonNull String getContentDescription() { return ""; }
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/mms/Slide.java]: Is there a reason not to make this abstract?

Comment (mcginty) [inline src/org/thoughtcrime/securesms/mms/Slide.java]: TextSlide didn't need to worry about it, but I can make it abstract and add an unused text description too if you'd prefer that.

Comment (moxie0) [inline src/org/thoughtcrime/securesms/mms/Slide.java]: no this is fine if there's one non-implementer and there's nothing catastrophic about the default behavior

**Comment by mcginty** [comment]
updated

File: src/org/thoughtcrime/securesms/components/TransferControlView.java [hunk 5]
```diff
@@ -65,12 +92,67 @@ public void showProgressSpinner() {
...
+      anim.addListener(new AnimatorListener() {
+        @Override public void onAnimationStart(Animator animation) {}
+        @Override public void onAnimationCancel(Animator animation) {}
+        @Override public void onAnimationRepeat(Animator animation) {}
+        @Override public void onAnimationEnd(Animator animation) {
+          ViewUtil.animateIn(view, inAnimation);
+        }
+      });
+      anim.start();
+    }
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/components/TransferControlView.java]: I think this method is probably too long, I can't follow what it does

**Comment by mcginty** [comment]
simplified view logic

**Pull request closed without merging**